src/Features/CSharp/Portable/CodeRefactorings/ConvertToRecord/ConvertToRecordHelpers.cs

```diff
@@ -0,0 +1,914 @@
+            // branches can either be block bodies or single statements
+            // each branch is followed by statements outside the branch either way
+            var trueOps = ((whenTrue as IBlockOperation)?.Operations ?? ImmutableArray.Create(whenTrue))
+                .Concat(otherOps);
+            var falseOps = ((whenFalse as IBlockOperation)?.Operations ??
+                (whenFalse != null
+                    ? ImmutableArray.Create(whenFalse)
+                    : ImmutableArray.Create<IOperation>()))
+                .Concat(otherOps);
+
+            // We expect one of the true or false branch to have exactly one statement: return false.
+            // If we don't find that, it indicates complex behavior such as
+            // extra statments, backup equality if one condition fails, or something else.
+            // We don't necessarily expect a return true because we could see
+            // a final return statement that checks a last set of conditions such as:
+            // if (other is C otherC)
+            // {
+            //     return otherC.A == A;
+            // }
+            // return false;
+            // We should never have a case where both branches could potentially return true;
+            // at least one branch must simply return false.
+            if (ReturnsFalseImmediately(trueOps) == ReturnsFalseImmediately(falseOps))
+                // both or neither fit the return false pattern, this if statement either doesn't do
+                // anything or does something too complex for us, so we assume it's not a default.
+                return false;
+
+            // when condition succeeds (evaluates to true), we return false
+            // so for equality the condition should not succeed
+            successRequirement = !ReturnsFalseImmediately(trueOps);
+            remainingStatements = successRequirement ? trueOps : falseOps;
+            remainingStatements.Concat(otherOps);
+            return true;
+        }
+
+        /// <summary>
+        /// Whether the equals method overrides object or IEquatable Equals method
+        /// </summary>
+        private static bool OverridesEquals(Compilation compilation, IMethodSymbol? equals, INamedTypeSymbol? equatableType)
+        {
+            var objectType = compilation.GetSpecialType(SpecialType.System_Object);
+            var objectEquals = objectType?.GetMembers(nameof(Equals)).FirstOrDefault() as IMethodSymbol;
+
+            if (equatableType != null && equals != null &&
+                equatableType.GetMembers(nameof(Equals)).FirstOrDefault() is IMethodSymbol equatableEquals &&
+                Equals(equals.ContainingType.FindImplementationForInterfaceMember(equatableEquals), equals))
+            {
+                return true;
+            }
+
+            while (equals != null)
+            {
+                if (Equals(objectEquals, equals))
```

Comment:
Ditto I'd expect objectEquals.Equals(equals) here since we don't guarantee reference equality of symbols, necessarily.